### PR TITLE
add LastEvent, hardwareType support and refactor

### PR DESCRIPTION
**Pull request by DarkPark**

### CODE REVIEW

File: src/commands/uplink/LastEvent.ts [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    isBatteryLow?: boolean,
+    /** there is a magnetic field influence */
+    isMagneticInfluence?: boolean,
+    /** button is release (device is unmounted) */
+    isButtonReleased?: boolean,
+    /** the device has detected a loss of connection to the server */
+    isConnectionLost?: boolean
+}
+
+interface ILastEvent2channelStatus {
```
Comment (DeadByElpy) [inline src/commands/uplink/LastEvent.ts:19]:
```suggestion
interface ILastEvent2ChannelStatus {
```
?

File: src/commands/uplink/LastEvent.ts [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    /** the battery voltage has dropped below the set threshold */
+    isBatteryLow?: boolean,
+    /** the device has detected a loss of connection to the server */
+    isConnectionLost?: boolean,
+    /** the first channel is not active */
+    isFirstChannelInactive?: boolean,
```
Comment (DeadByElpy) [inline src/commands/uplink/LastEvent.ts:25]:
```suggestion
    isFirstChannelActive?: boolean,
```
?

Comment (DarkPark) [reply src/commands/uplink/LastEvent.ts:25]: nope

`true` - the first channel is NOT active

File: src/utils/bitSet.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+export type TBitSetMask = Record<string, number>;
+
+export type TBooleanObject = Record<string, boolean>;
+
+
+/**
+ * Get integer number from object with boolean values values according to the given bit mask.
```
[the remark this answers is not in the document]
Comment (DarkPark) [reply src/utils/bitSet.ts:7]: these are internal methods but ok - added

File: src/commands/uplink/LastEvent.ts (hunk 1 above)
Comment (DarkPark) [reply src/commands/uplink/LastEvent.ts:19]: corrected

**Review by DeadByElpy — APPROVED** [review]

**Pull request merged**